frontend/src/editor/commands/addBorderNumbers.ts

```diff
@@ -17,114 +12,55 @@ function addBorderNumbers({ state, dispatch }: CommandProps): boolean {
   const schema = state.schema
 
   const paragraphNodeType = schema.nodes.paragraph
+  const numberNodeType = schema.nodes.borderNumberNumber
   const contentNodeType = schema.nodes.borderNumberContent
+  const borderNumberNodeType = schema.nodes.borderNumber
 
   let modified = false
-  let shift = 0
-  let updatedFrom = initialFrom
 
-  state.doc.nodesBetween(initialFrom, initialTo, (node, pos) => {
-    if (hasBorderNumber(state, pos, schema)) {
-      return
-    }
+  const nodesWithoutBorderNumbers: { node: ProsemirrorNode; pos: number }[] = []
+  state.doc.nodesBetween(initialFrom, initialTo, (node, pos, parent) => {
+    const isTopLevelNode = parent && parent.type === state.schema.topNodeType
 
-    if (isParagraphWithContent(node, paragraphNodeType)) {
-      const borderNumberNode = createBorderNumberNode(
-        node,
-        schema,
-        contentNodeType,
-      )
-      const currentPos = pos + shift
+    const isBorderNumberNode =
+      node.type === borderNumberNodeType || node.hasMarkup(borderNumberNodeType)
 
-      tr.replaceWith(currentPos, currentPos + node.nodeSize, borderNumberNode)
+    const isEmptyParagraph =
+      node.type === paragraphNodeType &&
+      node.childCount === 0 &&
+      node.textContent?.trim() === ""
 
-      const addedNodeSize = borderNumberNode.nodeSize - node.nodeSize
-      shift += addedNodeSize
+    if (!isTopLevelNode || isBorderNumberNode || isEmptyParagraph) return
 
-      const isFirstBorderNumber = pos < initialFrom
-      if (isFirstBorderNumber) {
-        /** This solution is not perfect. The tr.doc only contains the default "0".
-        The recalculated border numbers are not part of the tr.doc, hence the actual number of digits
-        (after the recalculation) is unknown and therefore the actual addedNodeSize. This might result in a
-         slightly shifted cursor position **/
-        updatedFrom = initialFrom + addedNodeSize
-      }
-
-      modified = true
-    }
+    nodesWithoutBorderNumbers.push({ node, pos })
   })
 
+  const reversedNodes = [...nodesWithoutBorderNumbers].reverse()
+  for (const { node, pos } of reversedNodes) {
+    const numberNode = numberNodeType.create(
+      null,
```

Comment:
Why do we need these null values while creating?

Reply:
If I remember correctly, I believe the first param are the attributes for the node. null would mean there are none. We don't have any for new ones, right?